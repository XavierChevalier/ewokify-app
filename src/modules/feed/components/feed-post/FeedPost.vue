<script setup lang="ts">
  import { computed, PropType } from 'vue'
  import { localeFormatDistance } from '@/tools/date/Date'
  import { Author } from '@/modules/feed/models/Author'
  import UserAvatar from '@/components/user-avatar/UserAvatar.vue'
  import LikeIcon from '@/components/like-icon/LikeIcon.vue'
  import CommentIcon from '@/components/comment-icon/CommentIcon.vue'

  const props = defineProps({
    author: {
      type: Object as PropType<Author>,
      required: true,
      // eslint-disable-next-line vue/valid-define-props
      validator: (value) => Author.safeParse(value).success,
    },
    content: {
      type: String,
      required: true,
    },
    numberOfLikes: {
      type: Number,
      required: true,
    },
    numberOfComments: {
      type: Number,
      required: true,
    },
    publicationDate: {
      type: Date,
      required: true,
    },
  })

  const publicationDateDistance = computed(() =>
    localeFormatDistance(props.publicationDate, new Date())
  )
</script>

<template>
  <div class="bg-white rounded p-5">
    <div class="flex items-center">
      <UserAvatar :user-name="author.name" :url="author.avatar" />

      <div class="flex flex-col pl-5">
        <span>{{ author.name }}</span>
        <span class="text-neutral">{{ publicationDateDistance }}</span>
      </div>
    </div>

    <p class="mt-5">{{ content }}</p>

    <div v-if="!!$slots.default" class="mt-2">
      <slot />
    </div>

    <div class="flex mt-5 justify-end">
      <LikeIcon show-counter :counter="numberOfLikes" />
      <CommentIcon show-counter :counter="numberOfComments" class="ml-5" />
    </div>
  </div>
</template>
