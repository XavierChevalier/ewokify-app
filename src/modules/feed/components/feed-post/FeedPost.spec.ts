import FeedPost from './FeedPost.vue'
import { generateStorybookSnapshotTests } from '@/tests/tools/StorybookSnapshot'
import {
  Default,
  WithTrack,
} from '@/modules/feed/components/feed-post/FeedPost.stories'

describe('FeedPost', () => {
  generateStorybookSnapshotTests({ Default, WithTrack })

  describe('author validation', () => {
    const validator = FeedPost.props.author?.validator
    it('should be defined', () => {
      expect(validator).toBeDefined()
    })

    describe.each`
      description                        | author
      ${'Empty author'}                  | ${{}}
      ${'Author with only a name'}       | ${{ name: 'TheBaronless' }}
      ${'Author with only an avatar'}    | ${{ avatar: 'https://picsum.photos/50' }}
      ${'Author with an integer avatar'} | ${{ avatar: 1 }}
      ${'Author with an non URL avatar'} | ${{ avatar: "it's just a string" }}
    `('$description', ({ author }) => {
      test(`GIVEN ANY invalid author state
        WHEN we validate it
        THEN it SHOULD ALWAYS be invalid
      `, () => {
        expect(validator(author)).toBe(false)
      })
    })

    describe.each`
      description       | author
      ${'Valid author'} | ${{ name: 'TheBaronless', avatar: 'https://picsum.photos/50' }}
    `('$description', ({ author }) => {
      test(`GIVEN ANY valid author state
        WHEN we validate it
        THEN it SHOULD ALWAYS be valid
      `, () => {
        expect(validator(author)).toBe(true)
      })
    })
  })
})
