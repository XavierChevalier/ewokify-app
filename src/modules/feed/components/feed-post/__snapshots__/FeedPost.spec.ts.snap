// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedPost should render the "Default" component correctly 1`] = `
<div
  class="bg-white rounded p-5"
>
  <div
    class="flex"
  >
    <div
      class="w-12 h-12 relative after:absolute after:w-14 after:h-14 after:-top-1 after:-left-1 after:rounded-full after:border-2 after:border-neutral"
    >
      <img
        alt="The TheBaronless's avatar"
        class="rounded-full w-full h-full"
        src="https://picsum.photos/50"
      />
    </div>
    <div
      class="flex flex-col pl-5"
    >
      <span>
        TheBaronless
      </span>
      <span
        class="text-neutral"
      >
        less than a minute ago
      </span>
    </div>
  </div>
  <p
    class="mt-5"
  >
    Lorem ipsum
  </p>
  
  
  <span
    class="flex mt-5 justify-end"
  >
    <svg
      class="text-neutral"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M12.1,18.55L12,18.65L11.89,18.55C7.14,14.24 4,11.39 4,8.5C4,6.5 5.5,5 7.5,5C9.04,5 10.54,6 11.07,7.36H12.93C13.46,6 14.96,5 16.5,5C18.5,5 20,6.5 20,8.5C20,11.39 16.86,14.24 12.1,18.55M16.5,3C14.76,3 13.09,3.81 12,5.08C10.91,3.81 9.24,3 7.5,3C4.42,3 2,5.41 2,8.5C2,12.27 5.4,15.36 10.55,20.03L12,21.35L13.45,20.03C18.6,15.36 22,12.27 22,8.5C22,5.41 19.58,3 16.5,3Z"
        style="fill: currentColor;"
      />
    </svg>
    <span
      class="ml-2"
    >
      10
    </span>
  </span>
</div>
`;

exports[`FeedPost should render the "WithTrack" component correctly 1`] = `
<div
  class="bg-white rounded p-5"
>
  <div
    class="flex"
  >
    <div
      class="w-12 h-12 relative after:absolute after:w-14 after:h-14 after:-top-1 after:-left-1 after:rounded-full after:border-2 after:border-neutral"
    >
      <img
        alt="The TheBaronless's avatar"
        class="rounded-full w-full h-full"
        src="https://picsum.photos/50"
      />
    </div>
    <div
      class="flex flex-col pl-5"
    >
      <span>
        TheBaronless
      </span>
      <span
        class="text-neutral"
      >
        less than a minute ago
      </span>
    </div>
  </div>
  <p
    class="mt-5"
  >
    Lorem ipsum
  </p>
  
  <strong>
    text
  </strong>
  
  <span
    class="flex mt-5 justify-end"
  >
    <svg
      class="text-neutral"
      height="24"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M12.1,18.55L12,18.65L11.89,18.55C7.14,14.24 4,11.39 4,8.5C4,6.5 5.5,5 7.5,5C9.04,5 10.54,6 11.07,7.36H12.93C13.46,6 14.96,5 16.5,5C18.5,5 20,6.5 20,8.5C20,11.39 16.86,14.24 12.1,18.55M16.5,3C14.76,3 13.09,3.81 12,5.08C10.91,3.81 9.24,3 7.5,3C4.42,3 2,5.41 2,8.5C2,12.27 5.4,15.36 10.55,20.03L12,21.35L13.45,20.03C18.6,15.36 22,12.27 22,8.5C22,5.41 19.58,3 16.5,3Z"
        style="fill: currentColor;"
      />
    </svg>
    <span
      class="ml-2"
    >
      10
    </span>
  </span>
</div>
`;
